tradeoffs.Rmd

```diff
@@ -0,0 +1,172 @@
+Such test duplication of external business logic may be hard to get right and maintain, and should rarely be necessary.
+
+["Broad" integration tests](https://martinfowler.com/bliki/IntegrationTest.html) are run against live services, and thus require no test doubles.
+
+Neither unit nor integration tests need to cover an API specification comprehensively.
+Testing the API is not your concern, but merely the narrow interface between the API and your code.
+As a rule of thumb, your code should handle, and your tests should cover the failure *modes* of the API, including HTTP response codes, API response codes or incorrect response formats.
+Perhaps you can additionally test some API requests against some known outputs, but this will always remain a sample.
+There's always a chance that some deeply nested JSON response will break your code.
+You may not be able to cover all such errors from an errand or changed API, and that's okay.
+
+## Example
+
+To illustrate these tradeoffs, consider the [biblids](http://subugoe.github.io/biblids/) package, which wraps the [DOI resolution proxy server REST API](https://www.doi.org/factsheets/DOIProxy.html#rest-api) at doi.org.
+Digital object identifiers (DOIs) are used to persistently identify scientific publications such as `10.1038/nature14236` for an article in the journal Nature.
+The doi.org system lets institutions register such (persistent) DOIs, and update (non-persistent) URLs where full texsts can currently be found.
+It might be helpful to query this API from R.
+
+Happily, the doi.org API is unauthenticated, highly performant and gives simple responses.
+For any given `GET` request, there are only these outcomes:
+
+1. A DOI cannot be found DOI.org (HTTP status code 402).
+2. A DOI can be found (HTTP status code 200).
+    a. The queried value can be found (API response code 1).
+    b. The queried value cannot be found (API response code 200).
+       For example, a DOI is registered but has no URL to resolve to.
+       Rare but theoretically possible.
+3. Something unspecified went wrong on the server end (HTTP status code 500).
+
+Omitting some bells and whistles from biblids, we can minimally wrap the API thus:
+
+```{r minimal-wrapper}
+# unexported helper
+# this is not tested, b/c the implementation could change
+verb_doi_handle <- function(doi, ...) {
+  httr::VERB(
+    url = "https://doi.org/",
+    # better to use curl::curl_escape() here
+    path = paste0("api/handles/", doi),
+    ...
+  )
+}
+
+# exported function
+# this is tested
+get_doi_handle <- function(doi, ...) {
+  resp <- verb_doi_handle(doi = doi, verb = "GET", ...)
+  httr::stop_for_status(resp) # covers outcomes 1 *and* 3 from above list
+  if (httr::content(resp)$responseCode == 200) {
+    # covers outcome 2b
+    stop("The handle exists, but the queried value(s) could not be found.")
```

Comment:
By the way one could imagine your function would not fail but return a NULL instead or some data.frame with some sort of explanatory column?

I can see the next function does something linked to that but it'd mean you call the API twice? Am I missing something?

Reply:
Not sure I fully understand.

Do you mean the function *should* return a `NULL` or `data.frame()` with some output if the queried value does not exist?

I haven't fully thought this through, but I like functions to fail early and to maintain type/length-stability at almost any cost, which is hard to do if you build these kind of special cases ...

But maybe I misunderstood your comment ...